File: src/helper/deposit_helper.ts
import { checkEthBalance, checkMNTBalance } from "./token_balance.js";
import {
  setupCrossChainMessengerFoEth,
  setupCrossChainMessengerForMnt,
} from "./cross_chain_messenger.js";
import { MessageStatus } from "@mantleio/sdk";
import { ethers } from "ethers";
import { environment } from "../utils/environment.js";
import { decimalStringToBigInt } from "../utils/maths.js";
import { ERROR_CODE, generateError, handleError } from "../utils/errors.js";

export async function depositETH(
  privateKey,
  amount
): Promise<{ totalExecutionTime; txnHash; ethBalanceOnL1; ethBalanceOnL2 }> {
  const start = new Date();

  const { l2Wallet, crossChainMessenger } = await setupCrossChainMessengerFoEth(
    privateKey
  );

  const eth = ethers.utils.parseEther(amount);
  let response;
  try {
    response = await crossChainMessenger.depositETH(eth, {
      overrides: { gasLimit: environment.ethGasLimit },
    });
  } catch (err) {
    const errorCode = handleError(err, ERROR_CODE.DEPOSIT_FAILED);
    throw generateError(errorCode);
  }

  console.log(`Deposit ETH transaction hash (on L1): ${response.hash}`);
  await response.wait();

  await crossChainMessenger.waitForMessageStatus(
    response,
    MessageStatus.RELAYED
  );

  const { l1Balance, l2Balance } = await checkEthBalance(
    l2Wallet,
    crossChainMessenger
  );

  return {
    totalExecutionTime: (new Date().getTime() - start.getTime()) / 1000,
    txnHash: response.hash,
    ethBalanceOnL1: l1Balance,
    ethBalanceOnL2: l2Balance,
  };
}

export async function depositMNT(
  privateKey,
  amount
): Promise<{ totalExecutionTime; txnHash; mntBalanceOnL1; mntBalanceOnL2 }> {
  const start = new Date();
  const { l1Wallet, crossChainMessenger, l1MntContract } =
    await setupCrossChainMessengerForMnt(privateKey);

  const precision = 18;
  const totalMNT = decimalStringToBigInt(amount, precision);

  let allowTxnResponse;
  try {
    allowTxnResponse = await crossChainMessenger.approveERC20(
      environment.l1Mnt,
      environment.l2Mnt,
      totalMNT.toString()
    );
  } catch (err) {
    const errorCode = handleError(err, ERROR_CODE.APPROVE_TXN_FAILED);
    throw generateError(errorCode);
  }
  await allowTxnResponse.wait();

  let response;
  try {
    response = await crossChainMessenger.depositMNT(totalMNT.toString(), {
      overrides: { gasLimit: environment.ethGasLimit },
    });
  } catch (err) {
    const errorCode = handleError(err, ERROR_CODE.DEPOSIT_FAILED);
    throw generateError(errorCode);
  }

  console.log(`Deposit MNT transaction hash (on L1): ${response.hash}`);
  await response.wait();

  await crossChainMessenger.waitForMessageStatus(
    response,
    MessageStatus.RELAYED
  );

  const { l1Balance, l2Balance } = await checkMNTBalance(
    l1Wallet,
    l1MntContract,
    crossChainMessenger
  );

  return {
    totalExecutionTime: (new Date().getTime() - start.getTime()) / 1000,
    txnHash: response.hash,
    mntBalanceOnL1: l1Balance,
    mntBalanceOnL2: l2Balance,
  };
}

File: src/internal/helper/deposit_helper.go
package helper

import (
	"context"
	"math/big"
	"time"

	"github.com/AbhinitKumarRai/rebalance-funds/src/pkg/utils"
	"github.com/ethereum/go-ethereum/accounts/abi/bind"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"
)

// DepositResponse represents the response from a deposit operation
type DepositResponse struct {
	TotalExecutionTime float64
	TxnHash            string
	L1Balance          float64
	L2Balance          float64
}

// DepositETH handles ETH deposits from L1 to L2
func DepositETH(privateKey string, amount float64) (*DepositResponse, error) {
	start := time.Now()

	// Setup cross-chain messenger
	setup, err := SetupCrossChainMessengerForEth(privateKey)
	if err != nil {
		return nil, err
	}

	// Convert amount to wei
	weiAmount := new(big.Int).Mul(
		big.NewInt(int64(amount*1e18)),
		big.NewInt(1e18),
	)

	// Create deposit transaction
	tx := types.NewTransaction(
		setup.L1Signer.Nonce.Uint64(),
		common.HexToAddress(utils.GetEnvironment().L2Bridge),
		weiAmount,
		uint64(utils.GetEnvironment().EthGasLimit),
		big.NewInt(0),
		nil,
	)

	// Sign and send transaction
	signedTx, err := setup.L1Signer.Signer(setup.L1Signer.From, tx)
	if err != nil {
		return nil, err
	}

	err = setup.L1Client.SendTransaction(context.Background(), signedTx)
	if err != nil {
		return nil, err
	}

	// Wait for transaction to be mined
	receipt, err := bind.WaitMined(context.Background(), setup.L1Client, signedTx)
	if err != nil {
		return nil, err
	}

	// Check balances
	l1Balance, l2Balance, err := CheckEthBalance(setup.L2Wallet, setup.L1Client, setup.L2Client)
	if err != nil {
		return nil, err
	}

	return &DepositResponse{
		TotalExecutionTime: time.Since(start).Seconds(),
		TxnHash:            receipt.TxHash.Hex(),
		L1Balance:          l1Balance,
		L2Balance:          l2Balance,
	}, nil
}

// DepositMNT handles MNT deposits from L1 to L2
func DepositMNT(privateKey string, amount float64) (*DepositResponse, error) {
	start := time.Now()

	// Setup cross-chain messenger
	setup, err := SetupCrossChainMessengerForMNT(privateKey)
	if err != nil {
		return nil, err
	}

	// Convert amount to wei
	weiAmount := new(big.Int).Mul(
		big.NewInt(int64(amount*1e18)),
		big.NewInt(1e18),
	)

	// Approve MNT transfer
	approveTx, err := setup.L1MNTContract.Transact(setup.L1Wallet, "approve", utils.GetEnvironment().L2Bridge, weiAmount)
	if err != nil {
		return nil, err
	}

	// Wait for approval
	_, err = bind.WaitMined(context.Background(), setup.L1Client, approveTx)
	if err != nil {
		return nil, err
	}

	// Create deposit transaction
	tx := types.NewTransaction(
		setup.L1Wallet.Nonce.Uint64(),
		common.HexToAddress(utils.GetEnvironment().L2Bridge),
		weiAmount,
		uint64(utils.GetEnvironment().EthGasLimit),
		big.NewInt(0),
		nil,
	)

	// Sign and send transaction
	signedTx, err := setup.L1Wallet.Signer(setup.L1Wallet.From, tx)
	if err != nil {
		return nil, err
	}

	err = setup.L1Client.SendTransaction(context.Background(), signedTx)
	if err != nil {
		return nil, err
	}

	// Wait for transaction to be mined
	receipt, err := bind.WaitMined(context.Background(), setup.L1Client, signedTx)
	if err != nil {
		return nil, err
	}

	// Check balances
	l1Balance, l2Balance, err := CheckMNTBalance(setup.L1Wallet, setup.L1MNTContract, setup.L2Client)
	if err != nil {
		return nil, err
	}

	return &DepositResponse{
		TotalExecutionTime: time.Since(start).Seconds(),
		TxnHash:            receipt.TxHash.Hex(),
		L1Balance:          l1Balance,
		L2Balance:          l2Balance,
	}, nil
}
